lib/modules/datasource/rubygems/index.ts

```diff
@@ -36,16 +34,35 @@ export class RubyGemsDatasource extends Datasource {
       /* eslint-disable @typescript-eslint/restrict-template-expressions */
       `${registryUrl}/${packageName}`,
   })
-  getReleases({
+  async getReleases({
     packageName,
     registryUrl,
   }: GetReleasesConfig): Promise<ReleaseResult | null> {
-    if (parseUrl(registryUrl)?.hostname === 'rubygems.org') {
-      return this.rubyGemsOrgDatasource.getReleases({ packageName });
+    // istanbul ignore if
+    if (!registryUrl) {
+      return null;
+    }
+
+    if (!this.versionsDatasources[registryUrl]) {
+      this.versionsDatasources[registryUrl] = new VersionsDatasource(
+        RubyGemsDatasource.id,
+        registryUrl
+      );
+    }
+
+    try {
+      return await this.versionsDatasources[registryUrl].getReleases({
```

Comment:
Will this try the versions approach for every single dependency even if it fails for all, or is there any caching to prevent duplicate tries/failures?

Reply:
There are 2 ways that this avoids excessive traffic:
- In `VersionsDatasource` I used the same convention that was there previously - `this.registryCache` - to cache for 15 minutes per one of our previous conversations.
- Since I have an indexed object in this file (`this.versionsDatasources[registryUrl]`), only a single instance gets created per registryUrl, so it shouldn't recheck per dependency - just once for the entire registry.

If you see something else I have missed in this equation, let me know.

Reply:
Perhaps to make this even more stable, `versionsDatasources` could be made a global and not an instance variable on `RubyGemsDatasource`, but I was working with the previous convention established by `this.rubyGemsOrgDatasource`... and since it appears there's only ever one instance of `RubyGemsDatasource` created (in `lib/modules/datasource/api.ts`), I would assume the way I have it would be fine.

Reply:
One difference is that we *knew* rubygems supports `/versions` so we don't face the prospect of "it keeps retrying even if no /versions exists".

Reply:
Right, except per my comments I'm not sure why it would keep retrying. For a given registry URL (NOT for each dependency), it checks once every 15 minutes and then caches the failure so it doesn't try again.

Reply:
I think I missed the 15 minute cache part. In addition though I'd like to make sure we handle it when we happen to look up e.g. 20 dependencies at once. i.e. we don't check `/version` in parallel 20 times due to the race condition

Reply:
See my other comment in `versions-datasource`.

Reply:
This code genericizes checking for `/versions` first and then falling back to `/api/v1` per the suggestion in https://github.com/renovatebot/renovate/issues/16126#issuecomment-1169645686 and elsewhere in that conversation. Are you suggesting adding an explicit "stop here and don't try the alternate" in the case that RubyGems returns a 404 on `/versions`?

Reply:
Added an exception